DatabaseManager' is not assignable to type v1.21.1

### 📜 Description

Type error databaseManager

 const databaseManager = DatabaseManager.fromConfig(config, {
    logger: root,
  });
const taskScheduler = TaskScheduler.fromConfig(config, { databaseManager });

### 👍 Expected behavior

we should not get type issue.
if we use following in packages/backend/src/index.ts. it works fine.
 const databaseManager:any = DatabaseManager.fromConfig(config, {
    logger: root,
  });

### 👎 Actual Behavior with Screenshots

Types have separate declarations of a private property 'config'.ts(2322)
(property) databaseManager?: DatabaseManager | undefined

### 👟 Reproduction steps

backstage version 1.20.0
1) Create a new backstage app 
2 yarn tsc
3) you will see typescript warning

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

I'm running into this issue where building with the multistage docker file fails, but `yarn tsc` locally outside of docker runs fine.

```
yarn tsc
yarn run v1.22.21
$ tsc
✨  Done in 1.96s.
```
```
docker image build -t backstage-multistage .
[+] Building 7.9s (18/24)                                                                                                                                               docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                    0.0s
 => => transferring dockerfile: 3.88kB                                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/library/node:18-bookworm-slim                                                                                                                1.3s
 => [internal] load .dockerignore                                                                                                                                                       0.0s
 => => transferring context: 141B                                                                                                                                                       0.0s
 => [packages 1/6] FROM docker.io/library/node:18-bookworm-slim@sha256:27c468a7089b35374ec79b2a49ae9911818e37665caffe4d4d531b74c78cfbe3                                                 0.0s
 => [internal] load build context                                                                                                                                                       0.0s
 => => transferring context: 30.88kB                                                                                                                                                    0.0s
 => CACHED [internal] settings cache mount permissions                                                                                                                                  0.0s
 => CACHED [build  2/10] RUN --mount=type=cache,target=/var/cache/apt,sharing=locked     --mount=type=cache,target=/var/lib/apt,sharing=locked     apt-get update &&     apt-get insta  0.0s
 => CACHED [build  3/10] RUN --mount=type=cache,target=/var/cache/apt,sharing=locked     --mount=type=cache,target=/var/lib/apt,sharing=locked     apt-get update &&     apt-get insta  0.0s
 => CACHED [build  4/10] WORKDIR /app                                                                                                                                                   0.0s
 => CACHED [packages 2/6] WORKDIR /app                                                                                                                                                  0.0s
 => CACHED [packages 3/6] COPY package.json yarn.lock ./                                                                                                                                0.0s
 => CACHED [packages 4/6] COPY packages packages                                                                                                                                        0.0s
 => CACHED [packages 5/6] COPY plugins plugins                                                                                                                                          0.0s
 => CACHED [packages 6/6] RUN find packages ! -name "package.json" -mindepth 2 -maxdepth 2 -exec rm -rf {} +                                                                            0.0s
 => CACHED [build  5/10] COPY --from=packages --chown=node:node /app .                                                                                                                  0.0s
 => CACHED [build  6/10] RUN --mount=type=cache,target=/home/node/.cache/yarn,sharing=locked,uid=1000,gid=1000     yarn install --frozen-lockfile --network-timeout 600000              0.0s
 => CACHED [build  7/10] COPY --chown=node:node . .                                                                                                                                     0.0s
 => ERROR [build  8/10] RUN yarn tsc                                                                                                                                                    6.5s
------                                                                                                                                                                                       
 > [build  8/10] RUN yarn tsc:                                                                                                                                                               
0.337 yarn run v1.22.19                                                                                                                                                                      
0.363 $ tsc                                                                                                                                                                                  
6.419 packages/backend/src/index.ts:43:60 - error TS2322: Type 'import("/app/packages/backend/node_modules/@backstage/backend-common/dist/index").DatabaseManager' is not assignable to type 'import("/app/node_modules/@backstage/backend-common/dist/index").DatabaseManager'.
6.419   Types have separate declarations of a private property 'config'.
6.419 
6.419 43   const taskScheduler = TaskScheduler.fromConfig(config, { databaseManager });
6.419                                                               ~~~~~~~~~~~~~~~
6.419 
6.419 
6.419 Found 1 error in packages/backend/src/index.ts:43
6.419 
6.448 error Command failed with exit code 2.
6.448 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
------
Dockerfile:41
--------------------
  39 |     COPY --chown=node:node . .
  40 |     
  41 | >>> RUN yarn tsc
  42 |     RUN yarn --cwd packages/backend build
  43 |     # If you have not yet migrated to package roles, use the following command instead:
--------------------
ERROR: failed to solve: process "/bin/sh -c yarn tsc" did not complete successfully: exit code: 2

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/vggbbybcq80umf6w60kpzhfr4
```

--- a/.changeset/brave-jobs-invite.md
+++ b/.changeset/brave-jobs-invite.md
@@ -1,0 +1,5 @@
+---
+'@backstage/backend-common': patch
+---
+
+Added a new `LegacyRootDatabaseService` interface that can be used to avoid direct dependencies on the `DatabaseManager`.

--- a/.changeset/lazy-dolls-roll.md
+++ b/.changeset/lazy-dolls-roll.md
@@ -1,0 +1,5 @@
+---
+'@backstage/backend-tasks': patch
+---
+
+The `TaskScheduler.fromConfig` method now accepts the `LegacyRootDatabaseService` interface rather than the full `DatabaseManager` implementation.

--- a/packages/backend-common/api-report.md
+++ b/packages/backend-common/api-report.md
@@ -249,7 +249,7 @@ export function createStatusCheckRouter(options: {
 }): Promise<express.Router>;
 
 // @public
-export class DatabaseManager {
+export class DatabaseManager implements LegacyRootDatabaseService {
   forPlugin(
     pluginId: string,
     deps?: {
@@ -554,6 +554,11 @@ export const legacyPlugin: (
     >;
   }>,
 ) => BackendFeature;
+
+// @public
+export type LegacyRootDatabaseService = {
+  forPlugin(pluginId: string): PluginDatabaseManager;
+};
 
 // @public
 export function loadBackendConfig(options: {

--- a/packages/backend-common/src/database/DatabaseManager.ts
+++ b/packages/backend-common/src/database/DatabaseManager.ts
@@ -30,6 +30,7 @@ import {
 import { PluginDatabaseManager } from './types';
 import path from 'path';
 import {
+  DatabaseService,
   LifecycleService,
   LoggerService,
   PluginMetadataService,
@@ -54,6 +55,14 @@ export type DatabaseManagerOptions = {
 };
 
 /**
+ * An interface that represents the legacy global DatabaseManager implementation.
+ * @public
+ */
+export type LegacyRootDatabaseService = {
+  forPlugin(pluginId: string): DatabaseService;
+};
+
+/**
  * Manages database connections for Backstage backend plugins.
  *
  * @public
@@ -65,7 +74,7 @@ export type DatabaseManagerOptions = {
  * set `prefix` which is used to prefix generated database names if config is
  * not provided.
  */
-export class DatabaseManager {
+export class DatabaseManager implements LegacyRootDatabaseService {
   /**
    * Creates a {@link DatabaseManager} from `backend.database` config.
    *

--- a/packages/backend-tasks/api-report.md
+++ b/packages/backend-tasks/api-report.md
@@ -4,10 +4,10 @@
 
 ```ts
 import { Config } from '@backstage/config';
-import { DatabaseManager } from '@backstage/backend-common';
 import { Duration } from 'luxon';
 import { HumanDuration as HumanDuration_2 } from '@backstage/types';
 import { JsonObject } from '@backstage/types';
+import { LegacyRootDatabaseService } from '@backstage/backend-common';
 import { Logger } from 'winston';
 import { PluginDatabaseManager } from '@backstage/backend-common';
 
@@ -83,7 +83,7 @@ export interface TaskScheduleDefinitionConfig {
 
 // @public
 export class TaskScheduler {
-  constructor(databaseManager: DatabaseManager, logger: Logger);
+  constructor(databaseManager: LegacyRootDatabaseService, logger: Logger);
   forPlugin(pluginId: string): PluginTaskScheduler;
   // (undocumented)
   static forPlugin(opts: {
@@ -95,7 +95,7 @@ export class TaskScheduler {
   static fromConfig(
     config: Config,
     options?: {
-      databaseManager?: DatabaseManager;
+      databaseManager?: LegacyRootDatabaseService;
       logger?: Logger;
     },
   ): TaskScheduler;

--- a/packages/backend-tasks/src/tasks/TaskScheduler.ts
+++ b/packages/backend-tasks/src/tasks/TaskScheduler.ts
@@ -17,6 +17,7 @@
 import {
   DatabaseManager,
   getRootLogger,
+  LegacyRootDatabaseService,
   PluginDatabaseManager,
 } from '@backstage/backend-common';
 import { Config } from '@backstage/config';
@@ -37,7 +38,7 @@ export class TaskScheduler {
   static fromConfig(
     config: Config,
     options?: {
-      databaseManager?: DatabaseManager;
+      databaseManager?: LegacyRootDatabaseService;
       logger?: Logger;
     },
   ): TaskScheduler {
@@ -50,7 +51,7 @@ export class TaskScheduler {
   }
 
   constructor(
-    private readonly databaseManager: DatabaseManager,
+    private readonly databaseManager: LegacyRootDatabaseService,
     private readonly logger: Logger,
   ) {}
 

